macros/src/query.rs

```diff
@@ -193,6 +203,327 @@ pub fn derive(input: DeriveInput) -> Result<TokenStream2> {
     })
 }
 
+fn derive_enum(enum_ident: Ident, vis: Visibility, data: DataEnum, lifetime: Lifetime) -> Result<TokenStream2> {
+    let mut dangling_variant = None;
+    let mut fetch_variants = TokenStream2::new();
+    let mut state_variants = TokenStream2::new();
+    let mut query_get_variants = TokenStream2::new();
+    let mut fetch_access_variants = TokenStream2::new();
+    let mut fetch_borrow_variants = TokenStream2::new();
+    let mut fetch_prepare_variants = TokenStream2::new();
+    let mut fetch_execute_variants = TokenStream2::new();
+    let mut fetch_release_variants = TokenStream2::new();
+    let mut fetch_for_each_borrow = TokenStream2::new();
+
+    for variant in &data.variants {
+        let (fields, queries): (Vec<syn::Member>, Vec<TokenStream2>) = match variant.fields {
+            syn::Fields::Named(ref fields) => fields
+                .named
+                .iter()
+                .map(|f| {
+                    (
+                        syn::Member::Named(f.ident.clone().unwrap()),
+                        query_ty(&lifetime, &f.ty),
+                    )
+                })
+                .unzip(),
+            syn::Fields::Unnamed(ref fields) => fields
+                .unnamed
+                .iter()
+                .enumerate()
+                .map(|(i, f)| {
+                    (
+                        syn::Member::Unnamed(syn::Index {
+                            index: i as u32,
+                            span: Span::call_site(),
+                        }),
+                        query_ty(&lifetime, &f.ty),
+                    )
+                })
+                .unzip(),
+            syn::Fields::Unit => continue,
+        };
+
+        let ident = variant.ident.clone();
+
+        if fields.is_empty() {
+            // Ignore empty variants since otherwise they would always match
+            continue;
+        }
+
+        let fetches = queries
+            .iter()
+            .map(|ty| quote! { <#ty as ::hecs::Query>::Fetch })
+            .collect::<Vec<_>>();
+
+        if dangling_variant.is_none() {
+            dangling_variant = Some(match variant.fields {
+                syn::Fields::Named(_) => quote! {
+                    Self::#ident {
+                        #(
+                            #fields: #fetches::dangling(),
+                        )*
+                    }
+                },
+                syn::Fields::Unnamed(_) => quote! {
+                    Self::#ident(#(#fetches::dangling()),*)
+                },
+                syn::Fields::Unit => unreachable!(),
+            });
+        }
+
+        fetch_variants.extend(match variant.fields {
+            syn::Fields::Named(_) => quote! {
+                #ident {
+                    #(
+                        #fields: #fetches,
+                    )*
+                },
+            },
+            syn::Fields::Unnamed(_) => quote! {
+                #ident(#(#fetches),*),
+            },
```

Comment:
Could we combine these two branches in every case by using the `{0: x, 1: y, ...}` constructor/pattern syntax for tuple variants? That might save a *lot* of duplication.

Reply:
It wasn't possible to combine every branch in that way.
Instead, I converted the State and Fetch enums to always be struct enums which eliminates almost all of the duplication.
[39209c6](https://github.com/Ralith/hecs/pull/384/commits/39209c68e04e53a37fb0e3ffbd35ecb985d71fa2)

Reply:
When matching, struct enum variants would use `Enum::StructVariant { a, b, c }` whereas tuple enum variants would use `Enum::EnumVariant {0: a, 1: b, 2: c}`. It probably would have been possible to generate a TokenStream called `match_args` but converting everything into a struct enum was less lines of code anyways.